### PR TITLE
Add cmake -O2 flag to match qmake optimizations

### PR DESCRIPTION
**Pull request by thehans**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -19,6 +19,7 @@ message(STATUS "Configuration: ${CONFIG_OPTIONS}")
 
 project(openscad)
 
+set(CMAKE_CXX_FLAGS "-O2")
```
Comment (kintel) [inline CMakeLists.txt]: This should be turned off when in debug mode. I thought cmake managed that automatically..

Comment (justbuchanan) [reply CMakeLists.txt]: If you set `CMAKE_CXX_FLAGS_RELEASE` instead, cmake will automatically merge your params into `CMAKE_CXX_FLAGS` based on what compilation mode is selected through `CMAKE_BUILD_TYPE`. See https://cmake.org/Wiki/CMake_Useful_Variables#Compilers_and_Tools.

Also note that setting a variable directly clobbers its previous value. What you probably want is to *append* "-O2", so you want something like `set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O2")`

**Comment by thehans** [comment]
I found that cmake does provide some default flags based on the CMAKE_BUILD_TYPE.  They can be seen in `CMakeCache.txt` after running `cmake .` The problem is that if no build type is given, then CMAKE_BUILD_TYPE its just an empty string, and cmake won't apply any of its 'builtin' flags.

So after defaulting to Release, I found that 4 additional tests(all related to issue 1672) are now failing.
        376 - cgalpngtest_issue1672 (Failed)
        621 - opencsgtest_issue1672 (Failed)
        863 - csgpngtest_issue1672 (Failed)
        1105 - throwntogethertest_issue1672 (Failed)

the ctest console output shows the following error for each of these:
`*** stack smashing detected ***: /home/hans/projects/openscad_test/openscad_cmake/tests/../openscad terminated`

I tried lowering the -O3 to -O2, -O1, and even removing it entirely, but I seem to still get this stack smashing error.  I use `make -j8 VERBOSE=1` to ensure that the changes make it into the Makefile.

**Comment by kintel** [comment]
Do you know where this stack smashing error message comes from?

**Comment by thehans** [comment]
Just got a `bt` from `gdb`

```
[New Thread 0x7fffd7239700 (LWP 12796)]
[Thread 0x7fffd7239700 (LWP 12796) exited]
*** stack smashing detected ***: /home/hans/projects/openscad/openscad terminated

Thread 1 "openscad" received signal SIGABRT, Aborted.
0x00007ffff2979428 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
54      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  0x00007ffff2979428 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007ffff297b02a in __GI_abort () at abort.c:89
#2  0x00007ffff29bb7ea in __libc_message (do_abort=do_abort@entry=1, fmt=fmt@entry=0x7ffff2ad345f "*** %s ***: %s terminated\n") at ../sysdeps/posix/libc_fatal.c:175
#3  0x00007ffff2a5d11c in __GI___fortify_fail (msg=<optimized out>, msg@entry=0x7ffff2ad3441 "stack smashing detected") at fortify_fail.c:37
#4  0x00007ffff2a5d0c0 in __stack_chk_fail () at stack_chk_fail.c:28
#5  0x000000000068fa7c in CGAL::convex_hull_3<std::_List_iterator<CGAL::Point_3<CGAL::Epick> >, CGAL::Polyhedron_3<CGAL::Epick, CGAL::Polyhedron_items_3, CGAL::HalfedgeDS_default, std::allocator<int> >, CGAL::Convex_hull_traits_3<CGAL::Epick, CGAL::Boolean_tag<true> > > (first=..., beyond=..., polyhedron=..., traits=...) at /usr/include/CGAL/convex_hull_3.h:941
#6  0x0000000000693692 in CGAL::convex_hull_3<std::_List_iterator<CGAL::Point_3<CGAL::Epick> >, CGAL::Polyhedron_3<CGAL::Epick, CGAL::Polyhedron_items_3, CGAL::HalfedgeDS_default, std::allocator<int> > > (polyhedron=..., beyond=..., 
    first=...) at /usr/include/CGAL/convex_hull_3.h:950
#7  CGALUtils::applyHull (children=..., result=...) at /home/hans/projects/openscad/src/cgalutils-applyops.cc:189
#8  0x000000000071a5aa in GeometryEvaluator::applyToChildren3D (this=this@entry=0x7fffffffaed0, node=..., op=op@entry=OpenSCADOperator::HULL) at /home/hans/projects/openscad/src/GeometryEvaluator.cc:113
#9  0x000000000071b1a6 in GeometryEvaluator::applyToChildren (this=this@entry=0x7fffffffaed0, node=..., op=op@entry=OpenSCADOperator::HULL) at /home/hans/projects/openscad/src/GeometryEvaluator.cc:96
#10 0x000000000071bdd8 in GeometryEvaluator::visit (this=0x7fffffffaed0, state=..., node=...) at /home/hans/projects/openscad/src/GeometryEvaluator.cc:1086
#11 0x00000000004aaf0e in NodeVisitor::traverse (this=this@entry=0x7fffffffaed0, node=..., state=...) at /home/hans/projects/openscad/src/NodeVisitor.cc:30
#12 0x000000000071968a in GeometryEvaluator::evaluateGeometry (this=0x7fffffffaed0, node=..., allownef=allownef@entry=false) at /home/hans/projects/openscad/src/GeometryEvaluator.cc:57
#13 0x00000000005a578d in CSGTreeEvaluator::visit (this=0x7fffffffafc0, state=..., node=...) at /home/hans/projects/openscad/src/CSGTreeEvaluator.cc:276
#14 0x00000000004aaf0e in NodeVisitor::traverse (this=this@entry=0x7fffffffafc0, node=..., state=...) at /home/hans/projects/openscad/src/NodeVisitor.cc:30
#15 0x00000000004aaf37 in NodeVisitor::traverse (this=this@entry=0x7fffffffafc0, node=..., state=...) at /home/hans/projects/openscad/src/NodeVisitor.cc:20
#16 0x00000000005a2b85 in CSGTreeEvaluator::buildCSGTree (this=this@entry=0x7fffffffafc0, node=...) at /home/hans/projects/openscad/src/CSGTreeEvaluator.cc:33
#17 0x0000000000742cce in MainWindow::compileCSG (this=this@entry=0xd847f0, procevents=<optimized out>) at /home/hans/projects/openscad/src/mainwin.cc:1163
#18 0x000000000074c62f in MainWindow::csgRender (this=0xd847f0) at /home/hans/projects/openscad/src/mainwin.cc:1872
#19 0x00000000007a60b6 in MainWindow::qt_static_metacall (_o=<optimized out>, _c=<optimized out>, _id=<optimized out>, _a=<optimized out>) at /home/hans/projects/openscad/moc_MainWindow.cpp:557
#20 0x00007ffff384084a in QMetaMethod::invoke(QObject*, Qt::ConnectionType, QGenericReturnArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument) const () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#21 0x00007ffff3845f0d in QMetaObject::invokeMethod(QObject*, char const*, Qt::ConnectionType, QGenericReturnArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument) () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#22 0x000000000074298d in QMetaObject::invokeMethod (val9=..., val8=..., val7=..., val6=..., val5=..., val4=..., val3=..., val2=..., val1=..., val0=..., member=<optimized out>, obj=0xd847f0)
    at /usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs.h:425
#23 MainWindow::compileDone (this=this@entry=0xd847f0, didchange=<optimized out>) at /home/hans/projects/openscad/src/mainwin.cc:1064
#24 0x000000000074b57e in MainWindow::compile (this=this@entry=0xd847f0, reload=reload@entry=false, forcedone=forcedone@entry=false) at /home/hans/projects/openscad/src/mainwin.cc:992
#25 0x000000000074baf9 in MainWindow::actionRenderPreview (this=0xd847f0) at /home/hans/projects/openscad/src/mainwin.cc:1861
#26 0x00000000007a60ac in MainWindow::qt_static_metacall (_o=<optimized out>, _c=<optimized out>, _id=<optimized out>, _a=<optimized out>) at /home/hans/projects/openscad/moc_MainWindow.cpp:556
#27 0x00007ffff3863d2a in QMetaObject::activate(QObject*, int, int, void**) () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#28 0x00007ffff411e412 in QAction::triggered(bool) () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#29 0x00007ffff4120898 in QAction::activate(QAction::ActionEvent) () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#30 0x00007ffff4121204 in QAction::event(QEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#31 0x00007ffff412805c in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#32 0x00007ffff412d516 in QApplication::notify(QObject*, QEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#33 0x00000000007590cf in OpenSCADApp::notify (this=<optimized out>, object=<optimized out>, event=<optimized out>) at /home/hans/projects/openscad/src/OpenSCADApp.cc:30
#34 0x00007ffff383538b in QCoreApplication::notifyInternal(QObject*, QEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#35 0x00007ffff3b9f888 in QShortcutMap::dispatchEvent(QKeyEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5
#36 0x00007ffff3b9f9c1 in QShortcutMap::tryShortcutEvent(QObject*, QKeyEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5
#37 0x00007ffff3b740ea in QGuiApplicationPrivate::processKeyEvent(QWindowSystemInterfacePrivate::KeyEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5
#38 0x00007ffff3b79185 in QGuiApplicationPrivate::processWindowSystemEvent(QWindowSystemInterfacePrivate::WindowSystemEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5
#39 0x00007ffff3b5cf08 in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5
#40 0x00007ffff7e86200 in ?? () from /usr/lib/x86_64-linux-gnu/libQt5XcbQpa.so.5
#41 0x00007ffff5ed4197 in g_main_context_dispatch () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#42 0x00007ffff5ed43f0 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#43 0x00007ffff5ed449c in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#44 0x00007ffff388b7cf in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#45 0x00007ffff3832b4a in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#46 0x00007ffff383abec in QCoreApplication::exec() () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#47 0x0000000000453f2d in gui (inputFiles=std::vector of length 1, capacity 1 = {...}, original_path=..., argc=argc@entry=1, argv=argv@entry=0x7fffffffdc08) at /home/hans/projects/openscad/src/openscad.cc:774
#48 0x0000000000442a0e in main (argc=1, argv=0x7fffffffdc08) at /home/hans/projects/openscad/src/openscad.cc:988
```

**Comment by thehans** [comment]
I did a lot more testing and realized it was the `-DNDEBUG` flag(cmake by default sets this flag for non-'Debug' build types) that was causing the additional tests to fail, not related to `-O` optimization.  So I overrode the cmake defaults to not set that flag.

I'm thinking I can open a separate issue for why `-DNDEBUG` causes `issue1672.scad` to crash.

I also temporarily used this PR to try out cmake on travis, but it doesn't find Eigen3 (#2082), so I've reverted those changes.

@kintel do you think these current changes are good to merge?

File: CMakeLists.txt [hunk 2]
```diff
@@ -19,6 +19,26 @@ message(STATUS "Configuration: ${CONFIG_OPTIONS}")
 
 project(openscad)
 
+# Default to Release build
+if(NOT CMAKE_BUILD_TYPE)
```
Comment (kintel) [inline CMakeLists.txt]: Usually, one shouldn't edit the CMAKE_BUILD_TYPE, as cmake may be configured to e.g. build both release and debug configurations at the same time. I have to look into how to best achieve an optimized default build.

File: CMakeLists.txt [hunk 3]
```diff
@@ -19,6 +19,26 @@ message(STATUS "Configuration: ${CONFIG_OPTIONS}")
...
+  message(STATUS "CMAKE_BUILD_TYPE not specified.  Defaulting to 'Release'")
+  message(STATUS "Usage: cmake -DCMAKE_BUILD_TYPE=[Debug|Release|RelWithDebInfo|MinSizeRel] .")
+  set(CMAKE_BUILD_TYPE Release)
+else()
+  message(STATUS "CMAKE_BUILD_TYPE: ${CMAKE_BUILD_TYPE}")
+endif()
+
+# cmake default flags include "-DNDEBUG" on all non-Debug build types, and does not put "-DDEBUG" on Debug builds
+# OpenSCAD does the opposite of that, so we override those build-type-specific flags here
+set(CMAKE_C_FLAGS_DEBUG          "-g -DDEBUG" CACHE STRING "Set Debug flags" FORCE)
```
Comment (kintel) [inline CMakeLists.txt]: All these settings sound a bit overkill as cmake already does this.
How about enabling `-DDEBUG` only for debug builds and let cmake handle the rest?
e.g. `add_compile_options("$<$<CONFIG:DEBUG>:-DDEBUG>")`

**Comment by kintel** [comment]
Did I understand correctly that you'd like to actively turn off `-DNDEBUG` due to an unknown error?

File: CMakeLists.txt (hunk 2 above)
Comment (thehans) [reply CMakeLists.txt]: As far as I understand it's only possible for build type to hold a single value.  You might be thinking about the 'RelWithDebInfo' (Release with debug info), build type regarding "both release and debug configurations at the same time"?  RelWithDebInfo is a default CMAKE_BUILD_TYPE that just specifies flags of: "-O2 -g -NDEBUG"

Also this code will only be set CMAKE_BUILD_TYPE if the user has not specified one.

**Comment by thehans** [comment]
@kintel Yes I am attempting to actively turn on `-DNDEBUG` due to an error, though I'm not sure I understand what you mean by it being "unknown"

Here's a summary of the facts:
- Adding -DNDEBUG flag to build (cmake default) causes additional tests to fail  
    - Only for tests related to issue1672.scad 
    - The bug results in "stack smashing detection" and program termination immediately upon preview
        - The backtrace from a crash has been logged
- existing build (qmake) does *not* ever set `-DNDEBUG` as far as I can tell
- `NDEBUG` occurs only a handful of times in the source, mostly as [hacks](https://github.com/openscad/openscad/blob/master/src/cgal.h#L10-L13) [to](https://github.com/openscad/openscad/blob/master/src/cgalutils-tess.cc#L5-L8) [disable](https://github.com/openscad/openscad/blob/master/src/polyset-utils-old.cc#L7-L10)
it.

File: CMakeLists.txt (hunk 2 above)
Comment (kintel) [reply CMakeLists.txt]: Try `cmake -G Xcode` or `cmake -g "Visual Studio 14 2015"`. That will generate both release and debug builds.

**Comment by kintel** [comment]
Sounds good, and you recent updates feel a lot cleaner!
By "unknown" I meant that the cause is still unknown, so I agree that we should log it and work around it in the interim.
I'm still uncertain about defaulting to release unless told otherwise. I'll try to research this a bit in the week-end unless someone beats me to it.

**Comment by kintel** [comment]
Looking into this: It looks like the cmake Xcode generator runs cmake multiple times setting `CMAKE_BUILD_TYPE` accordingly, so your default trick should work.

**Pull request merged**